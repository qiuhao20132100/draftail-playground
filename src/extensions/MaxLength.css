.ProgressMeter {
  overflow: visible;
  transform: rotate(-90deg);
}

@keyframes RadialCounterPulse {
  0% {
    stroke-width: 2px;
  }
  50% {
    stroke-width: 4px;
  }
  100% {
    stroke-width: 2px;
  }
}

.ProgressMeter__background {
  fill: none;
  stroke-width: 1px;
  stroke: #eeeeee;
}

.ProgressMeter__progressbar {
  fill: none;
  stroke-width: 2px;
}

.ProgressMeter--pulse .ProgressMeter__progressbar {
  animation: RadialCounterPulse 0.3s ease-in-out;
}
