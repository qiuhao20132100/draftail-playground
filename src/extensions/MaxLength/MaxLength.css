.overflow-mark {
  background-color: #ff4136;
}

.MaxLength__select {
  display: block;
  position: relative;
  border: 1px solid currentColor;
  background: transparent;
  height: 3rem;
  border-radius: 0;
}
