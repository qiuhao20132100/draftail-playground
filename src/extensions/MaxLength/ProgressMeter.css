.ProgressMeter {
  overflow: visible;
  transform: rotate(-90deg);
}

.ProgressMeter__background {
  fill: none;
  stroke-width: 1px;
  stroke: #eeeeee;
}

.ProgressMeter__progressbar {
  fill: none;
  stroke-width: 2px;
}

@keyframes RadialCounterPulse {
  0% {
    stroke-width: 2px;
  }
  50% {
    stroke-width: 4px;
  }
  100% {
    stroke-width: 2px;
  }
}

.ProgressMeter--pulse .ProgressMeter__progressbar {
  animation: RadialCounterPulse 0.3s ease-in-out;
}
