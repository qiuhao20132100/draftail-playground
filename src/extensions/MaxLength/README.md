# Max-length extension for Draftail

A Draftail extension to enforce a maximum length constraint on the editor content. The UI is inspired by Twitter’s character count, which highlights the text that goes over the tweets’ length limit.
