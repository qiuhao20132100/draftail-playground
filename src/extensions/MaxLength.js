import { EditorState } from "draft-js";
import PropTypes from "prop-types";
import React, { Component, PureComponent } from "react";
import { ToolbarButton } from "draftail";
import Select from "react-simpler-select";

import Modal from "../components/Modal";

import "./MaxLength.css";

const CONTENT_LENGTHS = {};
CONTENT_LENGTHS[140] = "Tweet";
CONTENT_LENGTHS[280] = "Double tweet";
CONTENT_LENGTHS[3 * 10 * 200] = "3-min read";
CONTENT_LENGTHS[211591 * 10 * 200] = "Crime and Punishment";

const CONTENT_LENGTH_OPTIONS = Object.keys(CONTENT_LENGTHS).map((value) => ({
  value,
  label: CONTENT_LENGTHS[value],
}));

const getDefaultThreshold = () => {
  return (
    JSON.parse(window.sessionStorage.getItem("threshold")) ||
    CONTENT_LENGTH_OPTIONS[0].value
  );
};

const getMeterColor = (progress) => {
  let color = "#1da1f2";

  if (progress >= 1) {
    color = "#ff4136";
  } else if (progress >= 0.9) {
    color = "orange";
  }

  return color;
};

const forceResetEditorState = (editorState) => {
  return EditorState.set(
    EditorState.createWithContent(
      editorState.getCurrentContent(),
      editorState.getDecorator(),
    ),
    {
      selection: editorState.getSelection(),
      undoStack: editorState.getUndoStack(),
      redoStack: editorState.getRedoStack(),
    },
  );
};

const delayAndIdle = (timeoutHandle, idleHandle, callback) => {
  if (timeoutHandle) {
    window.clearTimeout(timeoutHandle);
  }

  if (idleHandle) {
    window.clearIdleCallback(idleHandle);
  }

  if (window.requestIdleCallback) {
    timeoutHandle = window.setTimeout(() => {
      idleHandle = window.requestIdleCallback(callback, {
        timeout: 500,
      });
    }, 500);
  } else {
    timeoutHandle = window.setTimeout(callback, 1000);
  }
};

class ProgressMeter extends PureComponent {
  render() {
    const { radius, progress } = this.props;
    const diameter = radius * 2;
    const circumference = diameter * Math.PI;
    const isFull = progress >= 1;

    return (
      <svg
        height={diameter}
        width={diameter}
        className={`ProgressMeter Draftail-Icon${
          isFull ? " ProgressMeter--pulse" : ""
        }`}
      >
        <circle
          className="ProgressMeter__background"
          cx="50%"
          cy="50%"
          r={radius}
        />
        <circle
          className="ProgressMeter__progressbar"
          cx="50%"
          cy="50%"
          r={radius}
          stroke={getMeterColor(progress)}
          style={{
            strokeDasharray: circumference,
            strokeDashoffset: isFull
              ? 0
              : circumference - circumference * progress,
          }}
        />
      </svg>
    );
  }
}

/**
 * A basic control showing the reading time / content length for the editorâ€™s content.
 */
class MaxLength extends Component {
  constructor(props) {
    super(props);
    const threshold = getDefaultThreshold();

    this.state = {
      isOpen: false,
      threshold,
    };

    this.onRequestClose = this.onRequestClose.bind(this);
    this.onChangeThreshold = this.onChangeThreshold.bind(this);
    this.forceRenderDecorators = this.forceRenderDecorators.bind(this);
    this.delayForceResetEditorState = delayAndIdle.bind(
      null,
      this.timeoutHandle,
      this.idleHandle,
      this.forceRenderDecorators,
    );
  }

  onRequestClose() {
    this.setState({
      isOpen: false,
    });
  }

  onChangeThreshold(value) {
    const threshold = Number(value);
    this.setState({
      threshold,
    });

    window.sessionStorage.setItem("threshold", JSON.stringify(threshold));

    this.forceRenderDecorators();
  }

  forceRenderDecorators() {
    const { getEditorState, onChange } = this.props;
    const editorState = getEditorState();

    onChange(forceResetEditorState(editorState));
  }

  componentDidUpdate() {
    const { getEditorState } = this.props;
    const editorState = getEditorState();
    const lastChange = editorState.getLastChangeType();

    if (lastChange) {
      this.forceRenderDecorators();
    }
  }

  render() {
    const { getEditorState } = this.props;
    const { isOpen, threshold } = this.state;
    const editorState = getEditorState();
    const content = editorState.getCurrentContent();
    const contentLength = content
      .getBlockMap()
      .reduce((length, block) => length + block.getLength(), 0);

    return (
      <React.Fragment>
        <ToolbarButton
          name="MAX_LENGTH"
          title={`Max length: ${CONTENT_LENGTHS[threshold].label}`}
          icon={
            <ProgressMeter radius={8} progress={contentLength / threshold} />
          }
          onClick={() => {
            this.setState({
              isOpen: true,
            });
          }}
        />
        <Modal
          onRequestClose={this.onRequestClose}
          onAfterOpen={this.onAfterOpen}
          isOpen={isOpen}
          contentLabel="Max length chooser"
        >
          <label>
            Set a length threshold
            <Select
              value={threshold}
              options={CONTENT_LENGTH_OPTIONS}
              onChange={this.onChangeThreshold}
            />
          </label>
        </Modal>
      </React.Fragment>
    );
  }
}

MaxLength.propTypes = {
  getEditorState: PropTypes.func.isRequired,
};

export class MaxLengthDecorator {
  constructor() {
    this.blockLength = {};
    this.component = this.renderToken.bind(this);
    this.strategy = this.getDecorations.bind(this);
  }

  // Renders the decorated tokens.
  renderToken({ children }) {
    return <mark className="overflow-mark">{children}</mark>;
  }

  getDecorations(block, callback, contentState) {
    const blockKey = block.getKey();
    const blockLength = block.getLength();

    const previousContentLength = contentState
      .getBlockMap()
      .takeUntil((block) => block.getKey() === blockKey)
      .reduce((length, block) => length + block.getLength(), 0);

    const threshold = getDefaultThreshold();

    if (previousContentLength + blockLength > threshold) {
      console.log("decorate");
      const startOffset = Math.max(0, threshold - previousContentLength);
      callback(startOffset, blockLength);
    }
  }
}

export default MaxLength;
