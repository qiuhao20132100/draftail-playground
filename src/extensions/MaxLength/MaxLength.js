import { EditorState } from "draft-js";
import PropTypes from "prop-types";
import React, { Component } from "react";
import { ToolbarButton } from "draftail";
import Select from "react-simpler-select";

import Modal from "../../components/Modal";
import ProgressMeter from "./ProgressMeter";

import "./MaxLength.css";

const CONTENT_LENGTHS = {};
CONTENT_LENGTHS[140] = "Tweet";
CONTENT_LENGTHS[280] = "Double tweet";
CONTENT_LENGTHS[3 * 10 * 200] = "3-min read";
CONTENT_LENGTHS[211591 * 10 * 200] = "Crime and Punishment";

const CONTENT_LENGTH_OPTIONS = Object.keys(CONTENT_LENGTHS).map((value) => ({
  value,
  label: CONTENT_LENGTHS[value],
}));

const getDefaultThreshold = () => {
  return (
    JSON.parse(window.sessionStorage.getItem("threshold")) ||
    CONTENT_LENGTH_OPTIONS[0].value
  );
};

const forceResetEditorState = (editorState) => {
  return EditorState.set(
    EditorState.createWithContent(
      editorState.getCurrentContent(),
      editorState.getDecorator(),
    ),
    {
      selection: editorState.getSelection(),
      undoStack: editorState.getUndoStack(),
      redoStack: editorState.getRedoStack(),
    },
  );
};

const delayAndIdle = (timeoutHandle, idleHandle, callback) => {
  if (timeoutHandle) {
    window.clearTimeout(timeoutHandle);
  }

  if (idleHandle) {
    window.clearIdleCallback(idleHandle);
  }

  if (window.requestIdleCallback) {
    timeoutHandle = window.setTimeout(() => {
      idleHandle = window.requestIdleCallback(callback, {
        timeout: 500,
      });
    }, 500);
  } else {
    timeoutHandle = window.setTimeout(callback, 1000);
  }
};

/**
 * A basic control showing the reading time / content length for the editorâ€™s content.
 */
class MaxLength extends Component {
  constructor(props) {
    super(props);
    const threshold = getDefaultThreshold();

    this.state = {
      isOpen: false,
      threshold,
    };

    this.onRequestClose = this.onRequestClose.bind(this);
    this.onChangeThreshold = this.onChangeThreshold.bind(this);
    this.forceRenderDecorators = this.forceRenderDecorators.bind(this);
    this.delayForceResetEditorState = delayAndIdle.bind(
      null,
      this.timeoutHandle,
      this.idleHandle,
      this.forceRenderDecorators,
    );
  }

  onRequestClose() {
    this.setState({
      isOpen: false,
    });
  }

  onChangeThreshold(value) {
    const threshold = Number(value);
    this.setState({
      threshold,
    });

    window.sessionStorage.setItem("threshold", JSON.stringify(threshold));

    this.forceRenderDecorators();
  }

  forceRenderDecorators() {
    const { getEditorState, onChange } = this.props;
    const editorState = getEditorState();

    onChange(forceResetEditorState(editorState));
  }

  componentDidUpdate() {
    const { getEditorState } = this.props;
    const editorState = getEditorState();
    const lastChange = editorState.getLastChangeType();

    if (lastChange) {
      this.forceRenderDecorators();
    }
  }

  render() {
    const { getEditorState } = this.props;
    const { isOpen, threshold } = this.state;
    const editorState = getEditorState();
    const content = editorState.getCurrentContent();
    const contentLength = content
      .getBlockMap()
      .reduce((length, block) => length + block.getLength(), 0);

    return (
      <React.Fragment>
        <ToolbarButton
          name="MAX_LENGTH"
          title={`Max length: ${CONTENT_LENGTHS[threshold].label}`}
          icon={
            <ProgressMeter radius={8} progress={contentLength / threshold} />
          }
          onClick={() => {
            this.setState({
              isOpen: true,
            });
          }}
        />
        <Modal
          onRequestClose={this.onRequestClose}
          onAfterOpen={this.onAfterOpen}
          isOpen={isOpen}
          contentLabel="Maximum length"
        >
          <label>
            <p>How long should your content be?</p>
            <Select
              className="MaxLength__select"
              value={threshold}
              options={CONTENT_LENGTH_OPTIONS}
              onChange={this.onChangeThreshold}
            />
          </label>
        </Modal>
      </React.Fragment>
    );
  }
}

MaxLength.propTypes = {
  getEditorState: PropTypes.func.isRequired,
};

/**
 * Creates text decorations based on the length of text of each block, and its preceding blocks.
 */
export class MaxLengthDecorator {
  constructor() {
    this.blockLength = {};
    this.component = this.renderDecoration.bind(this);
    this.strategy = this.createDecorations.bind(this);
  }

  renderDecoration({ children }) {
    return <mark className="overflow-mark">{children}</mark>;
  }

  createDecorations(block, callback, contentState) {
    const blockKey = block.getKey();
    const blockLength = block.getLength();

    const previousContentLength = contentState
      .getBlockMap()
      .takeUntil((block) => block.getKey() === blockKey)
      .reduce((length, block) => length + block.getLength(), 0);

    const currentLength = previousContentLength + blockLength;
    const threshold = getDefaultThreshold();
    const shouldDecorate = currentLength > threshold;

    if (shouldDecorate) {
      // Decorate from the first character going over the threshold, to the end of the block.
      const startOffset = Math.max(0, threshold - previousContentLength);
      callback(startOffset, blockLength);
    }
  }
}

export default MaxLength;
