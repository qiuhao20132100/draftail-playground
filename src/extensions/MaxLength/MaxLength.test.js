import React from "react";
import { convertFromRaw } from "draft-js";
import { shallow } from "enzyme";
import MaxLength, { MaxLengthDecorator } from "./MaxLength";

describe("MaxLength", () => {
  it("works", () => {
    expect(
      shallow(<MaxLength radius={8} progress={0.5} />),
    ).toMatchInlineSnapshot();
  });
});
describe("MaxLengthDecorator", () => {
  it("decorates", () => {
    const decorator = new MaxLengthDecorator();
    expect(
      shallow(decorator.component(<div>Test!</div>)),
    ).toMatchInlineSnapshot();
  });

  describe("finds decorations", () => {
    it("single block below threshold", () => {
      const decorator = new MaxLengthDecorator();
      const callback = jest.fn();

      const content = convertFromRaw({
        entityMap: {},
        blocks: [
          {
            key: "a",
            text: "test",
          },
        ],
      });
      const block = content.getFirstBlock();
      decorator.strategy(block, callback, content);
      expect(callback).not.toHaveBeenCalledWith();
    });

    it("single block above threshold", () => {
      const decorator = new MaxLengthDecorator();
      const callback = jest.fn();

      const content = convertFromRaw({
        entityMap: {},
        blocks: [
          {
            key: "a",
            text: "test".repeat(200),
          },
        ],
      });
      const block = content.getFirstBlock();
      decorator.strategy(block, callback, content);
      expect(callback).not.toHaveBeenCalledWith();
    });

    it("multiple blocks below threshold", () => {
      const decorator = new MaxLengthDecorator();
      const callback = jest.fn();

      const content = convertFromRaw({
        entityMap: {},
        blocks: [
          {
            key: "a",
            text: "test",
          },
          {
            key: "b",
            text: "test",
          },
        ],
      });
      const block = content.getFirstBlock();
      decorator.strategy(block, callback, content);
      expect(callback).not.toHaveBeenCalledWith();
    });

    it("multiple blocks above threshold", () => {
      const decorator = new MaxLengthDecorator();
      const callback = jest.fn();

      const content = convertFromRaw({
        entityMap: {},
        blocks: [
          {
            key: "a",
            text: "test".repeat(200),
          },
          {
            key: "b",
            text: "test",
          },
        ],
      });
      const block = content.getFirstBlock();
      decorator.strategy(block, callback, content);
      expect(callback).not.toHaveBeenCalledWith();
    });
  });
});
