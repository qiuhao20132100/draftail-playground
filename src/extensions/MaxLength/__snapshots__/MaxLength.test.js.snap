// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaxLength works 1`] = `
<Fragment>
  <ToolbarButton
    active={false}
    icon={
      <ProgressMeter
        progress={NaN}
        radius={8}
      />
    }
    label={null}
    name="MAX_LENGTH"
    onClick={[Function]}
    title="Length: undefined"
  />
  <Modal
    contentLabel="Maximum length"
    isOpen={false}
    onRequestClose={[Function]}
  >
    <label>
      <p>
        How long should the content be?
      </p>
      <Select
        className="MaxLength__select"
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "Tweet",
              "value": "140",
            },
            Object {
              "label": "Double tweet",
              "value": "280",
            },
            Object {
              "label": "3-min read",
              "value": "6000",
            },
            Object {
              "label": "Crime and Punishment",
              "value": "423182000",
            },
          ]
        }
        placeholder={null}
        value={null}
      />
    </label>
  </Modal>
</Fragment>
`;
